### PR TITLE
Unique id abstract message id manager

### PR DESCRIPTION
**Pull request by quynhn**
Take the work of https://github.com/linagora/james-project/pull/484 and make it pass the "event" part of the messageIdManagerTests.

### CODE REVIEW

**Comment by chibenwa** [comment]
Please review : 

  - 03ab1ba44000fd52a2f0f1932554745d4451ff33 Introduce MessageIdMapper interface
  - 0b9b72048120757a0baf6985464093edf96bdd99 JAMES-1885: Wrong mailbox parameter verifying
  - 55f201a8b05b003537026adc573928cf9ab54764 JAMES-1885: Add more mock data for StoreQuota
  - 883553f707bdc2625d970461d52d61afb8a325fc JAMES-1885: Introduce the StoreMessageIdManager
  - 31b506839a3ee53b1632f955afdb9adfb25497db JAMES-1885: Make the SideEffect testing should be tested with StoreMessageIdManager

We found it hard to use mock (Genericity of the tests do not allow us to set up mocks without overriding each and every method). So we prefered implementing a naïve "memory" mapper for MessageIdMapper, and get the Side effect tests past on it. I guess these tests should disppear when we implement meory backend.

My comments about it : 

 - We need to create a ticket number for it
 - We need to merge #503 it's depending on
 - 0b9b72048120757a0baf6985464093edf96bdd99 and 55f201a8b05b003537026adc573928cf9ab54764 should be part of #503 
 - Squash 883553f707bdc2625d970461d52d61afb8a325fc and 31b506839a3ee53b1632f955afdb9adfb25497db together

**Comment by chibenwa** [comment]
You are missing some changes with the UID / MODSEQ stuff...

Cherry pick this before the StoreMessageIdManager commit : https://github.com/linagora/james-project/commit/f110b2303995dc57488378d7cbfcd1f344704025

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                mailboxSession);
+        }
+    }
+
+    protected MailboxMessage createMessage(Date internalDate, int size, int bodyStartOctet, SharedInputStream content, Flags flags, PropertyBuilder propertyBuilder, List<MessageAttachment> attachments, MailboxId mailboxId) throws MailboxException {
+        return new SimpleMailboxMessage(messageIdFactory.generate(), internalDate, size, bodyStartOctet, content, flags, propertyBuilder, mailboxId, attachments);
+    }
+    
+    private void dispatchFlagsChange(MailboxSession mailboxSession, MailboxId mailboxId, UpdatedFlags updatedFlags) throws MailboxException {
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxById(mailboxId);
```
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: Might be in the if closure.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageIdManagerTestSystem.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Override
+    public void clean() {
+        mapperFactory.clean();
+    }
+
+    private MailboxMessage createMessage(MailboxId mailboxId, Flags flags, MessageId messageId) {
+        MailboxMessage mailboxMessage = mock(MailboxMessage.class);
+        when(mailboxMessage.getMessageId()).thenReturn(messageId);
+        when(mailboxMessage.getUid()).thenReturn(UID);
```
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageIdManagerTestSystem.java]: All your messages will have the same UID

**Comment by quynhn** [comment]
```
All your messages will have the same UID
```
No, it does not. So I changed a bit to make it will be random value each time persist new Message

**Comment by quynhn** [comment]
```
Might be in the if closure.
```
Yes and I moved to if closure

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageIdManagerTestSystem.java [hunk 3]
```diff
@@ -59,9 +60,10 @@ public StoreMessageIdManagerTestSystem(MessageIdManager messageIdManager, Mailbo
     @Override
     public MessageId persist(MailboxId mailboxId, Flags flags) {
         MessageId messageId = messageIdFactory.generate();
+        MessageUid uid = MessageUid.of(new Random().nextLong());
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageIdManagerTestSystem.java]: We try to not put any random value in our test, to make them reproducible.
+ here you risk a collision with 2 identical messageUid

**Comment by chibenwa** [comment]
We lack tests : setInMailboxes should also remove from mailboxes.

We need to test accordingly in #503 and modify here the StoreMessageIdManager. #509 can be used as reference...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                .toSet();
+            HashSet<MailboxId> targetMailboxes = Sets.newHashSet(mailboxIds);
+            List<MailboxId> mailboxesToRemove = ImmutableList.copyOf(Sets.difference(currentMailboxes, targetMailboxes));
+
+            SetView<MailboxId> mailboxesToAdd = Sets.difference(targetMailboxes, currentMailboxes);
+
+            if (!mailboxesToAdd.isEmpty()) {
+                MailboxMessage mailboxMessage = mailboxMessages.get(0);
+
+                validateQuota(mailboxIds, mailboxSession, mailboxMessage);
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: Nope. You need to multiply by the number of new messages...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    
+    private void dispatchFlagsChange(MailboxSession mailboxSession, MailboxId mailboxId, UpdatedFlags updatedFlags) throws MailboxException {
+        if (updatedFlags.flagsChanged()) {
+            Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxById(mailboxId);
+            dispatcher.flagsUpdated(mailboxSession, updatedFlags.getUid(), mailbox, updatedFlags);
+        }
+    }
+
+    private void validateQuota(List<MailboxId> mailboxIds, MailboxSession mailboxSession, MailboxMessage mailboxMessage) throws MailboxException {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: You can easily add a times parameter here...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: You should also decrement the count you check for each deleted message.

I would write outside of the if : 

```
validateQuota(mailboxIds, mailboxSession, mailboxMessage, mailboxesToAdd.size() - mailboxesToRemove.size());

if (!mailboxesToAdd.empty()) {
    ...
}
if (!mailboxesToRemove.empty()) {
    ...
}
```

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 6]
```diff
@@ -212,6 +223,36 @@ public void setInMailboxesShouldNotChangeUidAndModSeqInOriginalMailbox() throws
     }
 
     @Test
+    public void setInMailboxesShouldAddedAndRemovedMailboxes() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: *AddAndRemoveFromMailboxes

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
test this please

**Pull request merged**